ros/src/computing/perception/localization/packages/ekf_localizer/src/ekf_localizer.cpp

```diff
@@ -0,0 +1,672 @@
+   *     [ 0, 0,                 0,                 0,             0,  1]
+   */
+
+  Eigen::MatrixXd X_curr(dim_x_, 1); // curent state
+  Eigen::MatrixXd X_next(dim_x_, 1); // predicted state
+  ekf_.getLatestX(X_curr);
+  DEBUG_PRINT_MAT(X_curr.transpose());
+
+  Eigen::MatrixXd P_curr;
+  ekf_.getLatestP(P_curr);
+
+  const int d_dim_x = dim_x_ex_ - dim_x_;
+  const double yaw = X_curr(IDX::YAW);
+  const double yaw_bias = X_curr(IDX::YAWB);
+  const double vx = X_curr(IDX::VX);
+  const double wz = X_curr(IDX::WZ);
+  const double dt = ekf_dt_;
+
+  /* Update for latest state */
+  X_next(IDX::X) = X_curr(IDX::X) + vx * cos(yaw + yaw_bias) * dt; // dx = v * cos(yaw)
+  X_next(IDX::Y) = X_curr(IDX::Y) + vx * sin(yaw + yaw_bias) * dt; // dy = v * sin(yaw)
+  X_next(IDX::YAW) = X_curr(IDX::YAW) + (wz)*dt;                   // dyaw = omega + omega_bias
+  X_next(IDX::YAWB) = yaw_bias;
+  X_next(IDX::VX) = vx;
+  X_next(IDX::WZ) = wz;
+
+  X_next(IDX::YAW) = std::atan2(std::sin(X_next(IDX::YAW)), std::cos(X_next(IDX::YAW)));
+
+
+  /* Set A matrix for latest state */
+  Eigen::MatrixXd A = Eigen::MatrixXd::Identity(dim_x_, dim_x_);
+  A(IDX::X, IDX::YAW) = -vx * sin(yaw + yaw_bias) * dt;
+  A(IDX::X, IDX::YAWB) = -vx * sin(yaw + yaw_bias) * dt;
+  A(IDX::X, IDX::VX) = cos(yaw + yaw_bias) * dt;
+  A(IDX::Y, IDX::YAW) = vx * cos(yaw + yaw_bias) * dt;
+  A(IDX::Y, IDX::YAWB) = vx * cos(yaw + yaw_bias) * dt;
+  A(IDX::Y, IDX::VX) = sin(yaw + yaw_bias) * dt;
+  A(IDX::YAW, IDX::WZ) = dt;
+
+  /* Set covariance matrix Q for process noise. Calc Q by velocity and yaw angle covariance :
+     dx = Ax + Jp*w -> Q = Jp*w_cov*Jp'          */
+  Eigen::MatrixXd Jp = Eigen::MatrixXd::Zero(2, 2); // coeff of deviation of vx & yaw
+  Jp << cos(yaw), -vx * sin(yaw),
+        sin(yaw), vx * cos(yaw);
+  Eigen::MatrixXd Q_vx_yaw = Eigen::MatrixXd::Zero(2, 2); // cov of vx and yaw
+  Q_vx_yaw(0, 0) = P_curr(IDX::VX, IDX::VX) * dt;         // covariance of vx - vx
+  Q_vx_yaw(1, 1) = P_curr(IDX::YAW, IDX::YAW) * dt;       // covariance of yaw - yaw
+  Q_vx_yaw(0, 1) = P_curr(IDX::VX, IDX::YAW) * dt;        // covariance of vx - yaw
+  Q_vx_yaw(1, 0) = P_curr(IDX::YAW, IDX::VX) * dt;        // covariance of yaw - vx
+
+  Eigen::MatrixXd Q = Eigen::MatrixXd::Zero(dim_x_, dim_x_);
+  Q.block(0, 0, 2, 2) = Jp * Q_vx_yaw * Jp.transpose(); // for pos_x & pos_y
+  Q(IDX::YAW, IDX::YAW) = cov_proc_yaw_d_;            // for yaw
+  Q(IDX::YAWB, IDX::YAWB) = cov_proc_yaw_bias_d_;     // for yaw bias
+  Q(IDX::VX, IDX::VX) = cov_proc_vx_d_;               // for vx
+  Q(IDX::WZ, IDX::WZ) = cov_proc_wz_d_;               // for wz
+
+  ekf_.predictWithDelay(X_next, A, Q);
+
+  // debug
+  Eigen::MatrixXd X_result(dim_x_, 1);
+  ekf_.getLatestX(X_result);
+  DEBUG_PRINT_MAT(X_result.transpose());
+  DEBUG_PRINT_MAT((X_result - X_curr).transpose());
+
+}
+
+/*
+ * measurementUpdatePose
+ */
+void EKFLocalizer::measurementUpdatePose(const geometry_msgs::PoseStamped &pose)
```

Comment:
this is another method you can test. 
1. What are the expected inputs? (both valid and invalid)
1. What is the expected behavior if an invalid pose is received?

Reply:
I added some test code in `fa92af3`. Also add chack for invalid values.